cupy/core/fusion.pyx

```diff
@@ -1,888 +1,104 @@
-        return '<_FusionVar {} {}-dim array>'.format(self.dtype, self.ndim)
-
-    def __iadd__(self, other):
-        return cupy.add(self, other, self)
-
-    def __isub__(self, other):
-        return cupy.subtract(self, other, self)
-
-    def __imul__(self, other):
-        return cupy.multiply(self, other, self)
-
-    def __idiv__(self, other):
-        return cupy.divide(self, other, self)
-
-    def __itruediv__(self, other):
-        return cupy.true_divide(self, other, self)
-
-    def __ifloordiv__(self, other):
-        return cupy.floor_divide(self, other, self)
-
-    def __imod__(self, other):
-        return cupy.remainder(self, other, self)
-
-    def __ipow__(self, other):
-        return cupy.power(self, other, self)
-
-    def __ilshift__(self, other):
-        return cupy.left_shift(self, other, self)
-
-    def __irshift__(self, other):
-        return cupy.right_shift(self, other, self)
-
-    def __iand__(self, other):
-        return cupy.bitwise_and(self, other, self)
-
-    def __ior__(self, other):
-        return cupy.bitwise_or(self, other, self)
-
-    def __ixor__(self, other):
-        return cupy.bitwise_xor(self, other, self)
-
-
-class _FusionHistory(object):
-
-    """History of operation exectuted in the target function of fusion.
-
-    Attributes:
-        preamble_set (set of str): The preambles of submodules.
-        submodules (dict from str to submodule): The submodules.
-        count (int): The number of variables in the fused function.
-
-        op_list (list of _FusionOp): The map operations.
-        param_list (list of _FusionVarCUDA): The parameters
-        local_list (list of _FusionVarCUDA): The local variables.
-
-    Only when fusing the reduction, the following attributes are updated.
-
-        reduce_op (tuple): One of the element of reduction.***._raws._ops.
-        reduce_identity (any type): The identity value of the reduction.
-        reduce_kwargs (dict or None): kwargs of the reduction.
-
-        premap_ret (_FusionVarCUDA or None): The target of reduction
-        postmap_param (_FusionVarCUDA or None): The result of reduction
-        postmap_op_list (list of FuisonOp): The post-map operations.
-        postmap_local_list (list of _FusionVarCUDA): The local variables which
-    appears in the post-map operations
-    """
-
-    def __init__(self):
-        self.preamble_set = set()
-        self.submodules = dict()
-        self.count = 0
+cdef tuple _fusion_argument_types = (
+    core.ndarray, numpy.ndarray, numpy.generic,
+    int, float, complex, bool, type(None))
 
-        self.op_list = []
-        self.param_list = []
-        self.local_list = []
 
-        self.reduce_op = None
-        self.reduce_identity = None
-        self.reduce_kwargs = None
+def _get_fused_kernel(name, func, args):
+    try:
+        _thread_local.is_fusing = True
+        kernel_info = _fusion_trace.trace(func, args)
+        kernel = _fusion_kernel.FusedKernel(name, *kernel_info)
```

Comment:
Shouldn't this line be out of `try... finally`?

Reply:
This line is executed not in kernel runtime. This line is also a part of fusion.